File: mobile/hybrid/RoutingEngine.kt
package com.nova.mobile.hybrid

import android.util.Log
import com.nova.mobile.command.BuiltInIntent

/**
 * RoutingEngine — The core decision maker.
 * Combines CapabilityResolver + DeviceDiscovery + RoutingPolicy
 * to decide where each request should be executed.
 */
class RoutingEngine(
    private val capabilityResolver: CapabilityResolver,
    private val deviceDiscovery: DeviceDiscovery,
    var policy: RoutingPolicy = RoutingPolicy.AUTOMATIC
) {
    companion object {
        private const val TAG = "RoutingEngine"
    }

    fun decide(intentName: String, rawText: String): ExecutionTarget {
        return when (policy) {
            RoutingPolicy.ALWAYS_LOCAL -> {
                Log.i(TAG, "Policy=ALWAYS_LOCAL → LOCAL_ANDROID")
                ExecutionTarget.LOCAL_ANDROID
            }

            RoutingPolicy.ALWAYS_NOVA_CORE -> {
                val target = if (deviceDiscovery.isNovaCoreOnline) {
                    ExecutionTarget.NOVA_CORE
                } else {
                    Log.w(TAG, "Policy=ALWAYS_NOVA_CORE but Core offline → fallback to LOCAL")
                    ExecutionTarget.LOCAL_ANDROID
                }
                target
            }

            RoutingPolicy.OFFLINE_MODE -> {
                Log.i(TAG, "Policy=OFFLINE_MODE → LOCAL_ANDROID enforced")
                ExecutionTarget.LOCAL_ANDROID
            }

            RoutingPolicy.FALLBACK -> {
                // Try Nova Core first, fallback to local if offline
                if (capabilityResolver.isNovaCoreRequired(rawText) && deviceDiscovery.isNovaCoreOnline) {
                    ExecutionTarget.NOVA_CORE
                } else {
                    ExecutionTarget.LOCAL_ANDROID
                }
            }

            RoutingPolicy.AUTOMATIC -> {
                val preferredTarget = capabilityResolver.resolveTarget(intentName, rawText)
                return when {
                    preferredTarget == ExecutionTarget.NOVA_CORE && deviceDiscovery.isNovaCoreOnline -> {
                        Log.i(TAG, "AUTOMATIC → NOVA_CORE (online, capability match)")
                        ExecutionTarget.NOVA_CORE
                    }
                    preferredTarget == ExecutionTarget.NOVA_CORE && !deviceDiscovery.isNovaCoreOnline -> {
                        Log.w(TAG, "AUTOMATIC → Core required but OFFLINE. Staying on LOCAL for degraded response.")
                        ExecutionTarget.LOCAL_ANDROID
                    }
                    else -> {
                        Log.i(TAG, "AUTOMATIC → LOCAL_ANDROID")
                        ExecutionTarget.LOCAL_ANDROID
                    }
                }
            }
        }
    }

    fun requiresNovaCoreButOffline(intentName: String, rawText: String): Boolean {
        val preferred = capabilityResolver.resolveTarget(intentName, rawText)
        return preferred == ExecutionTarget.NOVA_CORE && !deviceDiscovery.isNovaCoreOnline
    }
}

File: nova/mobile/hybrid/routing_engine.py
"""Routing Engine — Combines CapabilityResolver + DeviceDiscovery + RoutingPolicy."""

import logging
from nova.mobile.hybrid.strategy import ExecutionTarget, RoutingPolicy
from nova.mobile.hybrid.capability_resolver import CapabilityResolver
from nova.mobile.hybrid.device_discovery import DeviceDiscovery

logger = logging.getLogger("nova.mobile.hybrid.routing_engine")


class RoutingEngine:
    def __init__(
        self,
        capability_resolver: CapabilityResolver,
        device_discovery: DeviceDiscovery,
        policy: RoutingPolicy = RoutingPolicy.AUTOMATIC
    ):
        self.capability_resolver = capability_resolver
        self.device_discovery = device_discovery
        self.policy = policy

    def decide(self, intent_name: str, raw_text: str) -> ExecutionTarget:
        if self.policy == RoutingPolicy.ALWAYS_LOCAL:
            logger.info("Policy=ALWAYS_LOCAL → LOCAL_ANDROID")
            return ExecutionTarget.LOCAL_ANDROID

        if self.policy == RoutingPolicy.ALWAYS_NOVA_CORE:
            if self.device_discovery.is_nova_core_online:
                return ExecutionTarget.NOVA_CORE
            logger.warning("Policy=ALWAYS_NOVA_CORE but Core offline → fallback LOCAL")
            return ExecutionTarget.LOCAL_ANDROID

        if self.policy == RoutingPolicy.OFFLINE_MODE:
            logger.info("Policy=OFFLINE_MODE → LOCAL_ANDROID enforced")
            return ExecutionTarget.LOCAL_ANDROID

        if self.policy == RoutingPolicy.FALLBACK:
            if self.capability_resolver.is_nova_core_required(raw_text) and self.device_discovery.is_nova_core_online:
                return ExecutionTarget.NOVA_CORE
            return ExecutionTarget.LOCAL_ANDROID

        # AUTOMATIC
        preferred = self.capability_resolver.resolve_target(intent_name, raw_text)
        if preferred == ExecutionTarget.NOVA_CORE and self.device_discovery.is_nova_core_online:
            logger.info("AUTOMATIC → NOVA_CORE (online, capability match)")
            return ExecutionTarget.NOVA_CORE
        if preferred == ExecutionTarget.NOVA_CORE and not self.device_discovery.is_nova_core_online:
            logger.warning("AUTOMATIC → Core required but OFFLINE → degraded LOCAL response")
            return ExecutionTarget.LOCAL_ANDROID
        logger.info("AUTOMATIC → LOCAL_ANDROID")
        return ExecutionTarget.LOCAL_ANDROID

    def requires_nova_core_but_offline(self, intent_name: str, raw_text: str) -> bool:
        preferred = self.capability_resolver.resolve_target(intent_name, raw_text)
        return preferred == ExecutionTarget.NOVA_CORE and not self.device_discovery.is_nova_core_online
